### ISSUE TITLE
Does it matter what the shape of the observation space is (in terms of a Box object)?

### ISSUE THREAD

**Issue by chenzhiliang94** [openai/gym#1781]
Hi, I have a very simple question regarding how the Box object should be created when defining the observable space for a rl-agent.

Assume that the observable space is a 4-dimensional state. Does it matter if I defined the observable_space in the custom environment as:

`self.observation_space = spaces.Box(low=0, high=1, shape=(2,2), dtype=np.float16)`
where I have defined it as a 2x2 matrix.
**OR**
`self.observation_space = spaces.Box(low=0, high=1, shape=(1,4), dtype=np.float16)`
where I have defined it as a 1x4 matrix (or simply, a vector).

I feel this should be quite a straightforward question; does anyone know if such minor difference in defining the space will affect the RL-algorithm which I'm passing the env into?

**Comment by MouseAndKeyboard** [comment]
Could you clarify what you mean?
Does your observation space have 4 features, for example:
1. Brightness,
2. Speed,
3. Time,
4. Colour

Or are you actually modelling a 4 dimensional space? 

If you're just doing the first thing all you need to do is define a dictionary:
https://github.com/openai/gym/blob/master/gym/spaces/dict.py
and have each feature be a record in the dictionary.

If you have an actual 4D environment for example a (2x2x2x2) environment `spaces.Box` can have any number of dimensions:
`.observation_space = spaces.Box(low=0, high=1, shape=(2,2,2,2), dtype=np.float16)`

**Comment by chenzhiliang94** [comment]
I mean, isn't a shape = (2,2) environment equivalent to having 4 features? Am I misunderstanding here?

**Comment by MouseAndKeyboard** [comment]
You should use the "box" environment object when 1 of your features is a spacial area.
For example:
one of my features was

1. a 32x64 screen of a game,
2. one feature for the score of the game (directly taken out of the game as an integer)
3. one feature for the time which has elapsed since the start of the game

Your observation space may be:
```
Dictionary {
    screen: Box(shape=(32x64)),
    score: Discrete(1),
    time_elapsed: Discrete(1)
}
```
the 32x64 box could be thought of as an array of length 32 where each element is an array of 64.

so to answer your question directly:

> what is the difference between a (2,2) space and a (1,4) 

(2,2) would look like this:
```
[
   [1,2],
   [3,4]
]
```
whereas (1,4) might look like this:
```
[
   [1,2,3,4]
]
```

**Comment by chenzhiliang94** [comment]
Hi,

Ah, my question is more like, if I have a feature taking on continuous value, it would appear as something like this in the dictionary: `'velocity': spaces.Box(low=-1, high=1, shape=(1,))` ? This is similar to a simple 1-dimensional continuous feature.

My initial question was more like, is there a difference in the RL learning algorithm if I pass in my features into a box, or into other kinds of formats? (Assuming all taking on continuous variables) Will putting all my features into a box somehow make the training process instable?

Zhi Liang

**Comment by thomasw-mitutoyo-ctl** [comment]
> Your observation space may be:
> 
> ```
> Dictionary {
>     screen: Box(shape=(32,64)),
>     score: Discrete(1),
>     time_elapsed: Discrete(1)
> }
> ```

Is it correct that the Python syntax for this is

    from gym.spaces import Box, Dict, Discrete
    env = Dict(screen=Box(shape=(32,64)), score=Discrete(1), time_elapsed=Discrete(1))

**Comment by christopherhesse** [comment]
I think this depends on your training setup, but I would not expect it to matter.

Convolutions will take advantage of the structure for something like screen pixels, so the shape matters there.

For a handful of unrelated generic real valued inputs, it's likely that this will be flattened to a 1-dimensional vector anyway, and in that case the structure won't matter at all.

**Comment by chenzhiliang94** [comment]
@christopherhesse 
Thanks for the reply! This is the straightforward answer I am looking for!

**Comment by baha2r** [comment]
Hi,

I still have this question does the system take advantage of having a label for its obs-space?
Or more simply, do we favor the system by creating a Dict space instead of Box?

obs= Dict(position=Box(shape=(3,1)), vel=Box(shape=(3,1)) 
VS 
obs= Box(low, high, shape=(6,1))

**Comment by pseudo-rnd-thoughts** [comment]
Depends what system you are talking about, it would be easier to go with option 1 for humans however neural network or machine learning systems don't care what you call the variables or what order they are in. 
In addition, a lot of reinforcement learning training systems are not designed to use `Dict` spaces and need to be flattened so it will most like turned into option 2, no matter what
